Spigot-Server-Patches/0235-Anti-Xray.patch

```diff
@@ -0,0 +1,1477 @@
++    public int a(PacketDataSerializer packetdataserializer, Chunk chunk, boolean flag, int i, PacketPlayOutMapChunkInfo packetPlayOutMapChunkInfo) { // Paper - Anti-Xray - Add parameter for packet info
+         int j = 0;
+         ChunkSection[] achunksection = chunk.getSections();
+         int k = 0;
+@@ -107,7 +143,7 @@ public class PacketPlayOutMapChunk implements Packet<PacketListenerPlayOut> {
+ 
+             if (chunksection != Chunk.a && (!this.e() || !chunksection.a()) && (i & 1 << k) != 0) {
+                 j |= 1 << k;
+-                chunksection.getBlocks().b(packetdataserializer);
++                chunksection.getBlocks().writeBlocks(packetdataserializer, packetPlayOutMapChunkInfo, k); // Paper - Anti-Xray
+                 packetdataserializer.writeBytes(chunksection.getEmittedLightArray().asBytes());
+                 if (flag) {
+                     packetdataserializer.writeBytes(chunksection.getSkyLightArray().asBytes());
+diff --git a/src/main/java/net/minecraft/server/PlayerChunk.java b/src/main/java/net/minecraft/server/PlayerChunk.java
+index 48a008e0..2eea3c59 100644
+--- a/src/main/java/net/minecraft/server/PlayerChunk.java
++++ b/src/main/java/net/minecraft/server/PlayerChunk.java
+@@ -134,6 +134,8 @@ public class PlayerChunk {
+             return false;
+         } else if (!this.chunk.isReady()) {
+             return false;
++        } else if (!this.chunk.world.chunkPacketBlockController.onChunkPacketCreate(this.chunk, '\uffff', false)) { // Paper - Anti-Xray
++            return false; // Paper - Anti-Xray - Wait and try again later
+         } else {
+             this.dirtyCount = 0;
+             this.h = 0;
+@@ -154,6 +156,7 @@ public class PlayerChunk {
+ 
+     public void sendChunk(EntityPlayer entityplayer) {
+         if (this.done) {
++            this.chunk.world.chunkPacketBlockController.onChunkPacketCreate(this.chunk, '\uffff', true); // Paper - Anti-Xray
+             entityplayer.playerConnection.sendPacket(new PacketPlayOutMapChunk(this.chunk, '\uffff'));
+             this.playerChunkMap.getWorld().getTracker().a(entityplayer, this.chunk);
+         }
+@@ -218,6 +221,7 @@ public class PlayerChunk {
+                         this.a(this.playerChunkMap.getWorld().getTileEntity(blockposition));
+                     }
+                 } else if (this.dirtyCount == 64) {
++                    this.chunk.world.chunkPacketBlockController.onChunkPacketCreate(this.chunk, this.h, true); // Paper - Anti-Xray
+                     this.a((Packet) (new PacketPlayOutMapChunk(this.chunk, this.h)));
+                 } else {
+                     this.a((Packet) (new PacketPlayOutMultiBlockChange(this.dirtyCount, this.dirtyBlocks, this.chunk)));
+diff --git a/src/main/java/net/minecraft/
```

Comment:
The existence of INSTANCE confused me at first, but then it made sense. I think renaming INSTANCE to NO_OP would be clearer of its intent.